File: sdk-python/iwf/tests/workflows/state_options_workflow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from iwf.command_request import CommandRequest
from iwf.command_results import CommandResults
from iwf.communication import Communication

from iwf.iwf_api.models import RetryPolicy, WaitUntilApiFailurePolicy
from iwf.persistence import Persistence
from iwf.state_decision import StateDecision
from iwf.state_schema import StateSchema
from iwf.workflow import ObjectWorkflow
from iwf.workflow_context import WorkflowContext
from iwf.workflow_state import T, WorkflowState
from iwf.workflow_state_options import WorkflowStateOptions


class InitState1(WorkflowState[str]):
    def get_state_options(self) -> WorkflowStateOptions:
        return WorkflowStateOptions(
            wait_until_api_retry_policy=RetryPolicy(maximum_attempts=1),
            proceed_to_execute_when_wait_until_retry_exhausted=WaitUntilApiFailurePolicy.PROCEED_ON_FAILURE,
        )

    def wait_until(
        self,
        ctx: WorkflowContext,
        input: T,
        persistence: Persistence,
        communication: Communication,
    ) -> CommandRequest:
        raise RuntimeError("test failure")

    def execute(
        self,
        ctx: WorkflowContext,
        input: T,
        command_results: CommandResults,
        persistence: Persistence,
        communication: Communication,
    ) -> StateDecision:
        data = "InitState1_execute_completed"
        return StateDecision.graceful_complete_workflow(data)


class InitState2(WorkflowState[str]):
    def get_state_options(self) -> WorkflowStateOptions:
        return WorkflowStateOptions(
            wait_until_api_retry_policy=RetryPolicy(maximum_attempts=1),
            proceed_to_execute_when_wait_until_retry_exhausted=WaitUntilApiFailurePolicy.FAIL_WORKFLOW_ON_FAILURE,
        )

    def wait_until(
        self,
        ctx: WorkflowContext,
        input: T,
        persistence: Persistence,
        communication: Communication,
    ) -> CommandRequest:
        raise RuntimeError("test failure")

    def execute(
        self,
        ctx: WorkflowContext,
        input: T,
        command_results: CommandResults,
        persistence: Persistence,
        communication: Communication,
    ) -> StateDecision:
        data = "InitState2_execute_completed"
        return StateDecision.graceful_complete_workflow(data)


class StateOptionsWorkflow1(ObjectWorkflow):
    def get_workflow_states(self) -> StateSchema:
        return StateSchema.with_starting_state(InitState1())


class StateOptionsWorkflow2(ObjectWorkflow):
    def get_workflow_states(self) -> StateSchema:
        return StateSchema.with_starting_state(InitState2())

File: sdk-java/src/test/java/io/iworkflow/integ/stateoptions/StateOptionsWorkflow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.iworkflow.integ.stateoptions;

import io.iworkflow.core.Context;
import io.iworkflow.core.ObjectWorkflow;
import io.iworkflow.core.StateDecision;
import io.iworkflow.core.StateDef;
import io.iworkflow.core.WorkflowState;
import io.iworkflow.core.WorkflowStateOptions;
import io.iworkflow.core.command.CommandRequest;
import io.iworkflow.core.command.CommandResults;
import io.iworkflow.core.communication.Communication;
import io.iworkflow.core.persistence.DataAttributeDef;
import io.iworkflow.core.persistence.Persistence;
import io.iworkflow.core.persistence.PersistenceFieldDef;
import io.iworkflow.gen.models.PersistenceLoadingPolicy;
import io.iworkflow.gen.models.PersistenceLoadingType;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@Component
public class StateOptionsWorkflow implements ObjectWorkflow {
    public static final String DA_WAIT_UNTIL = "DA_WAIT_UNTIL";
    public static final String DA_EXECUTE = "DA_EXECUTE";
    public static final String DA_BOTH = "DA_BOTH";

    @Override
    public List<PersistenceFieldDef> getPersistenceSchema() {
        return Arrays.asList(
                DataAttributeDef.create(String.class, DA_WAIT_UNTIL),
                DataAttributeDef.create(String.class, DA_EXECUTE),
                DataAttributeDef.create(String.class, DA_BOTH)
        );
    }

    @Override
    public List<StateDef> getWorkflowStates() {
        return Arrays.asList(
                StateDef.startingState(new StateOptionsWorkflowState1()),
                StateDef.nonStartingState(new StateOptionsWorkflowState2()),
                StateDef.nonStartingState(new StateOptionsWorkflowState3())
        );
    }
}

class StateOptionsWorkflowState1 implements WorkflowState<Void> {
    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest waitUntil(final Context context, final Void input, Persistence persistence, final Communication communication) {
        return CommandRequest.empty;
    }

    @Override
    public StateDecision execute(final Context context, final Void input, final CommandResults commandResults, Persistence persistence, final Communication communication) {
        persistence.setDataAttribute(StateOptionsWorkflow.DA_EXECUTE, "execute");
        persistence.setDataAttribute(StateOptionsWorkflow.DA_WAIT_UNTIL, "wait_until");
        persistence.setDataAttribute(StateOptionsWorkflow.DA_BOTH, "both");

        return StateDecision.singleNextState(StateOptionsWorkflowState2.class);
    }
}

class StateOptionsWorkflowState2 implements WorkflowState<Void> {
    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest waitUntil(final Context context, final Void input, Persistence persistence, final Communication communication) {
        final String daWaitUntil = persistence.getDataAttribute(StateOptionsWorkflow.DA_WAIT_UNTIL, String.class);

        if (!daWaitUntil.equals("wait_until")) {
            throw new RuntimeException("Expected DA_WAIT_UNTIL to be 'wait_until', got " + daWaitUntil);
        }

        final String daExecute = persistence.getDataAttribute(StateOptionsWorkflow.DA_EXECUTE, String.class);
        if (daExecute != null) {
            throw new RuntimeException("Expected DA_EXECUTE to be null, got " + daExecute);
        }

        return CommandRequest.empty;
    }

    @Override
    public StateDecision execute(final Context context, final Void input, final CommandResults commandResults, Persistence persistence, final Communication communication) {
        final String daExecute = persistence.getDataAttribute(StateOptionsWorkflow.DA_EXECUTE, String.class);

        if (!daExecute.equals("execute")) {
            throw new RuntimeException("Expected DA_EXECUTE to be 'execute', got " + daExecute);
        }

        final String daWaitUntil = persistence.getDataAttribute(StateOptionsWorkflow.DA_WAIT_UNTIL, String.class);
        if (daWaitUntil != null) {
            throw new RuntimeException("Expected DA_WAIT_UNTIL to be null, got " + daWaitUntil);
        }

        return StateDecision.singleNextState(StateOptionsWorkflowState3.class);
    }

    @Override
    public WorkflowStateOptions getStateOptions() {
        return new WorkflowStateOptions()
                .setWaitUntilApiDataAttributesLoadingPolicy(
                        new PersistenceLoadingPolicy()
                                .persistenceLoadingType(PersistenceLoadingType.PARTIAL_WITH_EXCLUSIVE_LOCK)
                                .partialLoadingKeys(Collections.singletonList(StateOptionsWorkflow.DA_WAIT_UNTIL))
                )
                .setExecuteApiDataAttributesLoadingPolicy(
                        new PersistenceLoadingPolicy()
                                .persistenceLoadingType(PersistenceLoadingType.PARTIAL_WITH_EXCLUSIVE_LOCK)
                                .partialLoadingKeys(Collections.singletonList(StateOptionsWorkflow.DA_EXECUTE)));
    }
}

class StateOptionsWorkflowState3 implements WorkflowState<Void> {
    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest waitUntil(final Context context, final Void input, Persistence persistence, final Communication communication) {
        final String str = persistence.getDataAttribute(StateOptionsWorkflow.DA_BOTH, String.class);

        if (!str.equals("both")) {
            throw new RuntimeException("Expected DA_BOTH to be 'both', got " + str);
        }

        return CommandRequest.empty;
    }

    @Override
    public StateDecision execute(final Context context, final Void input, final CommandResults commandResults, Persistence persistence, final Communication communication) {
        final String str = persistence.getDataAttribute(StateOptionsWorkflow.DA_BOTH, String.class);

        if (!str.equals("both")) {
            throw new RuntimeException("Expected DA_BOTH to be 'both', got " + str);
        }

        return StateDecision.gracefulCompleteWorkflow("success");
    }

    @Override
    public WorkflowStateOptions getStateOptions() {
        return new WorkflowStateOptions()
                .setDataAttributesLoadingPolicy(
                        new PersistenceLoadingPolicy()
                                .persistenceLoadingType(PersistenceLoadingType.PARTIAL_WITH_EXCLUSIVE_LOCK)
                                .partialLoadingKeys(Collections.singletonList(StateOptionsWorkflow.DA_BOTH))
                );
    }
}